031-statefulset-removal.md

```diff
@@ -0,0 +1,231 @@
+* The Pods will be always created in the same namespace as the node set
+* Garbage collection will be used to delete the pods
+
+But instead of using a Pod templates and generating Pods with fixed names, it will use full Pod structures including the custom names.
+It tries to merge the good from StatefulSet and eliminated the limitations we identified.
+
+The KafkaNodeSet will be created by the `KafkaAssemblyOperator` class in the same place in the reconciliation as the StatefulSet.
+This is also where the node sets will be patched when configuration changes.
+
+A new controller will be developed to reconcile the KafkaNodeSet resources.
+It will run as part of the Strimzi cluster operator container.
+It will follow more lightweight controller approach rather than the long-running _heavy_ approach of our current assembly operators.
+It will instead be designed to iterate quickly rather than wait.
+So it will for example not wait for Pod readiness.
+It will consume events about the KafkaNodeSets and their Pods and react to them.
+It will run in its own thread and use a work queue to queue the events and the reconciliations.
+
+![Controller design](images/031-kafkanodeset-controller.png)
+
+It will be responsible for:
+* Creating pods which do not exist
+* Deleting old pods removed from the KafkaNodeSet
+* Update the KafkaNodeSet status subresource
+
+When a new KafkaNodeSet is created, it will create its pods.
+When a KafkaNodeSet is modified, it will check if any pods need to be added or deleted.
+When a Pod is deleted or modified, it will identify the KafkaNodeSet it belongs to and reconcile the whole KafkaNodeSet.
+
+![KafkaNodeSet hierarchy](images/031-strimzi-with-kafkanodeset.png)
+
+The KafkaNodeSet controller will not be responsible for handling the rolling updates.
+They will continue to be handled in the same way as today by the `KafkaAssemblyOperator` and in the accompanied classes such as `KafkaRoller` and `ZooKeeperRoller`.
+They will identify which pods need to be rolled and delete them.
+The only difference will be that instead of the Kubernetes StatefulSet controller, it will be the KafkaNodeSet controller which will create a new pod.
+For each Pod managed by a KafkaNodeSet, Strimzi will calculate a revision (really a _hash_ of the Pod configuration).
+This revision will be used to identify changes to the desired Pod configuration and trigger rolling of the pod.
+This is a similar process as used by the StatefulSet itself.
+
+Similarly to the StatefulSets, the KafkaNodeSets will be stable resources as well.
+All the annotations currently used on StatefulSets will be applied to them.
+This applies to the annotations used to store the configurations and state but also to the command annotations such as manual rolling update triggers.
+
+### Feature Gate
+
+The switch from StatefulSets to KafkaNodeSets should be controller by a new feature gate called `StatefulSetRemoval`.
```

Comment:
What do you think to do the other way around `KafkaNodeSetEnable` which is false by default? I see the removal as the "negative" logic I don't like, but anyway I am fine to stick with it, was just a suggestion.

Reply:
^^ I personally prefer this approach - good suggestion.